sdk/websocket.go

```diff
@@ -73,24 +66,146 @@ type SubscribeHash struct {
 	Ch chan *HashInfo
 }
 
+func (s *SubscribeTransaction) Unsubscribe() error {
+	return s.subscribe.unsubscribe()
+}
+
+func (s *SubscribeHash) Unsubscribe() error {
+	return s.subscribe.unsubscribe()
+}
+
 type SubscribePartialRemoved struct {
 	*subscribe
 	Ch chan *PartialRemovedInfo
 }
 
+func (s *SubscribePartialRemoved) Unsubscribe() error {
+	return s.subscribe.unsubscribe()
+}
+
 type SubscribeStatus struct {
 	*subscribe
 	Ch chan *StatusInfo
 }
 
+func (s *SubscribeStatus) Unsubscribe() error {
+	return s.subscribe.unsubscribe()
+}
+
 type SubscribeSigner struct {
 	*subscribe
 	Ch chan *SignerInfo
 }
 
-type SubscribeError struct {
-	*subscribe
-	Ch chan *ErrorInfo
+func (s *SubscribeSigner) Unsubscribe() error {
+	return s.subscribe.unsubscribe()
+}
+
+type sendJson struct {
+	Uid       string `json:"uid"`
+	Subscribe string `json:"subscribe"`
+}
+
+func (c *ClientWebsocket) changeURLPort() {
+	c.config.BaseURL.Scheme = "ws"
+	c.config.BaseURL.Path = "/ws"
+	split := strings.Split(c.config.BaseURL.Host, ":")
+	host, port := split[0], "3000"
+	c.config.BaseURL.Host = strings.Join([]string{host, port}, ":")
+}
+
+func NewConnectWs(host string, timeout time.Duration) (*ClientWebsocket, error) {
+	u, err := url.Parse(host)
+	if err != nil {
+		return nil, err
+	}
+	newconf := &Config{BaseURL: u}
+	c := &ClientWebsocket{config: newconf}
+	c.common.client = c
+	c.Subscribe = (*SubscribeService)(&c.common)
+	c.timeout = time.Now().Add(timeout * time.Millisecond)
+
+	err = c.wsConnect()
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (c *ClientWebsocket) buildSubscribe(destination string) *subscribe {
+	b := new(subscribe)
+	b.Uid = c.Uid
+	b.Subscribe = destination
+	b.conn = c.client
+	return b
+}
+
+func (c *ClientWebsocket) wsConnect() error {
+	c.changeURLPort()
+	conn, err := websocket.Dial(c.config.BaseURL.String(), "", "http://localhost")
+	if err != nil {
+		return err
+	}
+	c.client = conn
+
+	conn.SetDeadline(c.timeout)
+
+	var msg []byte
+	if err = websocket.Message.Receive(c.client, &msg); err != nil {
+		return err
+	}
+
+	imsg, err := msgParser(msg)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	if err != nil {
+		return err
+	}
+	c.Uid = imsg.Uid
+
+	return nil
+}
+
+func (c *ClientWebsocket) subsChannel(s *subscribe) error {
+	if err := websocket.JSON.Send(c.client, sendJson{
+		Uid:       s.Uid,
+		Subscribe: s.Subscribe,
+	}); err != nil {
+		return err
+	}
+
+	var e error
```

Comment:
Name value is too short.
and is there a need to make the first variable 'err' local?

Reply:
It came from merge. Not my changes.